import{s as w,h as T}from"./app-78262cf9.js";const C={},D=()=>C,E=Object.freeze({cancelLastRequest:!0,cancelOnDispose:!0,useLastRequest:!1,cacheTime:0,retryTimes:0}),L=(b,h)=>{const e={...E,...D(),...h},f=w(null),m=w(!1),v=w(null),o={result:f,loading:m,error:v};let u=new AbortController,t=!1,d=0,c=null,s=0;const p=r=>{var n,l;o.result.value=r,o.error.value=null,a(!1),d=Date.now(),c=r,(n=e.onSuccess)==null||n.call(e,r),(l=e.onAfter)==null||l.call(e)},q=r=>{var n,l;r.message==="canceled"||r.name==="AbortError"||(o.result.value=null,o.error.value=r,a(!1),(n=e.onError)==null||n.call(e,r),(l=e.onAfter)==null||l.call(e))},a=r=>{o.loading.value=r,t=r},g=async(...r)=>{const n=u;for(;s<=e.retryTimes;){try{const l=await b(n.signal,...r);return n.signal.aborted?new Error("canceled"):(p(l),l)}catch(l){if(l.message==="canceled")return l;if(s===e.retryTimes)return q(l),l}s++}return new Error("unexpected error")},A=async(...r)=>{var n;return e.useLastRequest&&t?new Error("waiting for last request"):((n=e.onBefore)==null||n.call(e),s=0,c&&Date.now()-d<e.cacheTime?(a(!0),await new Promise(l=>setTimeout(()=>{var y;(y=e.onCache)==null||y.call(e,c),l(c)},20)),c):(e.cancelLastRequest&&t&&(i(),u=new AbortController),a(!0),await g(...r)))},R=async(...r)=>{var n;return(n=e.onBefore)==null||n.call(e),s=0,t&&(i(),u=new AbortController),a(!0),await g(...r)},i=()=>{var r;(r=e.onCancel)==null||r.call(e),a(!1),u.abort("cancel request")};return T(()=>{e.cancelOnDispose&&t&&i()}),{result:f,loading:m,error:v,run:A,forceRun:R,cancel:i}},O=L;export{O as F};
